src/core/socket-bind.c

```diff
@@ -0,0 +1,205 @@
+                return 0;
+        }
+
+        r = prepare_socket_bind_bpf(/*unit=*/NULL, /*allow_rules=*/NULL, /*deny_rules=*/NULL, &obj);
+        if (r < 0) {
+                log_debug_errno(r, "BPF based socket_bind is not supported: %m");
+                return 0;
+        }
+
+        return can_link_bpf_program(obj->progs.socket_bind_v4);
+}
+
+int socket_bind_install(Unit *u) {
+        _cleanup_(bpf_link_freep) struct bpf_link *ipv4 = NULL, *ipv6 = NULL;
+        _cleanup_(socket_bind_bpf_freep) struct socket_bind_bpf *obj = NULL;
+        _cleanup_free_ char *cgroup_path = NULL;
+        _cleanup_close_ int cgroup_fd = -1;
+        CGroupContext *cc;
+        int r;
+
+        cc = unit_get_cgroup_context(u);
+        if (!cc)
+                return 0;
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, u->cgroup_path, NULL, &cgroup_path);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to get cgroup path: %m");
+
+        if (!cc->socket_bind_allow && !cc->socket_bind_deny)
+                return 0;
+
+        r = prepare_socket_bind_bpf(u, cc->socket_bind_allow, cc->socket_bind_deny, &obj);
+        if (r < 0)
+                return log_unit_error_errno(u, r, "Failed to load BPF object: %m");
+
+        cgroup_fd = open(cgroup_path, O_RDONLY | O_CLOEXEC, 0);
+        if (cgroup_fd < 0)
+                return log_unit_error_errno(
+                                u, errno, "Failed to open cgroup=%s for reading", cgroup_path);
+
+        ipv4 = bpf_program__attach_cgroup(obj->progs.socket_bind_v4, cgroup_fd);
+        r = libbpf_get_error(ipv4);
+        if (r != 0)
+                return log_unit_error_errno(u, r, "Failed to link '%s' cgroup-bpf program",
+                                bpf_program__name(obj->progs.socket_bind_v4));
+
+        ipv6 = bpf_program__attach_cgroup(obj->progs.socket_bind_v6, cgroup_fd);
+        r = libbpf_get_error(ipv6);
+        if (r != 0)
+                return log_unit_error_errno(u, r, "Failed to link '%s' cgroup-bpf program",
+                                bpf_program__name(obj->progs.socket_bind_v6));
+
+        if (!u->socket_bind) {
+                u->socket_bind = new(SocketBind, 1);
+                if (!u->socket_bind)
+                        return log_oom();
+                *u->socket_bind = (SocketBind) {};
+        }
+
+        u->socket_bind->ipv4_link = TAKE_PTR(ipv4);
+        u->socket_bind->ipv6_link = TAKE_PTR(ipv6);
+
+        return 0;
+}
+#else /* ! BPF_FRAMEWORK */
+SocketBind *socket_bind_free(SocketBind *socket_bind) {
+        if (!socket_bind)
+                return NULL;
+
+        return mfree(socket_bind);
+}
+
+int socket_bind_supported(void) {
+        return 0;
+}
+
+int socket_bind_install(Unit *u) {
+         log_unit_debug_errno(u, SYNTHETIC_ERRNO(EOPNOTSUPP),
+                         "Failed to install AllowBind: BPF programs built from source code are not supported: %m");
```

Comment:
Strictly speaking,  SYNTHETIC_ERRNO() means that the code is not used for anything except the return value. And here the return value is not used…